dissect/hypervisor/backup/vbk.py

```diff
@@ -0,0 +1,1044 @@
+        """Get an entry from the vector.
+
+        Args:
+            idx: The index of the entry to get.
+        """
+        if idx >= self.count:
+            raise IndexError("MetaVector2 index out of range")
+        return self.type.from_bytes(self.vbk, self.data(idx))
+
+
+class MetaVector2(MetaVector[T]):
+    """A vector of meta items in the VBK file. Version 2.
+
+    Args:
+        vbk: The VBK object that the vector is part of.
+        type_: The type of the items in the vector.
+        page: The page number of the first page in the vector.
+        count: The number of items in the vector.
+
+    References:
+        - CMetaVec2
+    """
+
+    # MetaVector2 is essentially a table of page numbers that contain the vector entries
+    # The table pages of a MetaVector2 have a 8-16 byte header, so we can hold a maximum of 510-511 entries per page
+    # Read the comments in _lookup_page for more information
+    MAX_TABLE_ENTRIES_PER_PAGE = (PAGE_SIZE - 8) // 8
+
+    def __init__(self, vbk: VBK, type_: type[T], page: int, count: int):
+        super().__init__(vbk, type_, page, count)
+
+        # It's not actually a meta blob, but the same mechanism is used (next page pointer in the header)
+        # The table itself is essentially a big array of 64 bit integers, so cast it to a memoryview of that
+        self._table = xmemoryview(self.vbk.read_meta_blob(page), "<q")
+        self._lookup_page = lru_cache(128)(self._lookup_page)
+
+    def _lookup_page(self, idx: int) -> int:
+        """Look up the page number for an entry in the vector.
+
+        Args:
+            idx: The page index to lookup the page number for.
+        """
+        page_id = idx + 1
+
+        # MetaVec2 pages are a little special
+        # The first page has a 16 byte header:
+        # - 8 bytes for the next page number
+        # - 8 bytes for the root page number
+        # The second page has a 32 byte header:
+        # - 8 bytes for the next page number
+        # - 8 bytes for the second page number (this page)
+        # - 8 bytes for the third page number
+        # - 8 bytes for the fourth page number
+        # The third and fourth pages only have a 8 byte header:
+        # - 8 bytes for the next page number
+        # The fifth page has a 32 byte header again containing the next 3 page numbers
+        # We've not seen a table large enough to see this repeat more than once, but presumably it does
+        #
+        # This means that the first page can hold 510 entries, the second 508, and the third and fourth 511 each
+        # Reverse engineering gives us this funny looking loop, but it works, so use it for now
+        if page_id > self.MAX_TABLE_ENTRIES_PER_PAGE - 1:
+            page_id_offset = 1
+            while True:
+                prev_page_id_offset = page_id_offset
+                page_id_offset *= 4
+
+                if (4 * (self.MAX_TABLE_ENTRIES_PER_PAGE - 1)) * prev_page_id_offset >= page_id:
```

Comment:
Not available as a unit test file since this code path is only hit on very large files. There are probably some files available in the lab that hit this.

What would you suggest doing here instead?

Reply:
I've changed the logic a bit, let me know what you think. It's a bit simplistic but it works 😄